topi/include/topi/nn/pooling.h

```diff
@@ -285,6 +286,120 @@ inline Tensor global_pool(const Tensor& x,
   }
 }
 
+inline Expr start_index(Expr index, Expr odim, Expr idim) {
+  return tvm::floor((index * idim) / odim);
+}
+
+inline Expr end_index(Expr index, Expr odim, Expr idim) {
+  return tvm::ceil((index + 1) * idim / odim);
+}
+
+/*!
+* \brief Perform adaptive pooling on height and width dimension of data.
+*
+* \param x The input tensor
+* \param output_size Vector of two ints: {output_height, output_width}
+* \param pool_type The type of pooling operator
+* \param height_axis index of the height dimension
+* \param width_axis index of the width dimension
+*
+* \return The output tensor in same layout order
+*/
+inline Tensor adaptive_pool_impl(const Tensor& x,
+                                 const Array<Expr>& output_size,
+                                 PoolType pool_type,
+                                 const size_t height_axis,
+                                 const size_t width_axis) {
+  CHECK(x->shape.size() >= 2) << "Adaptive Pooling input must >= 2-D (H, W)";
+  CHECK_EQ(output_size.size(), 2) << "Adaptive Pooling output_size must have 2 elements";
+
+  auto height = x->shape[height_axis];
+  auto width = x->shape[width_axis];
+  auto out_height = tvm::ir::Simplify(output_size[0]);
+  auto out_width = tvm::ir::Simplify(output_size[1]);
+
+  Array<Expr> out_shape = x->shape;
+  out_shape.Set(height_axis, out_height);
+  out_shape.Set(width_axis, out_width);
+
+  if (pool_type == kMaxPool) {
+    return tvm::compute(out_shape, [&](const Array<Var>& output_indices) {
+      Array<Expr> indices;
+      for (const Var& var : output_indices)
+        indices.push_back(var);
+      auto height_index = indices[height_axis];
+      auto h_start = start_index(height_index, out_height, height);
+      auto h_end = end_index(height_index, out_height, height);
+      auto dheight = tvm::reduce_axis(Range(h_start, h_end));
```

Comment:
Does `tvm::reduce_axis` have to be called outside of `tvm::compute`? Here's the schedule for regular pooling:
```
// attr [tensor] storage_scope = "global"
allocate tensor[float32 * 33792]
produce tensor {
  parallel (ax0.ax1.fused, 0, 256) {
    for (ax2, 0, 12) {
      for (ax3, 0, 11) {
        tensor[((((ax0.ax1.fused*12) + ax2)*11) + ax3)] = -340282346638528859811704183484516925440.000000f
        for (rv, 0, 3) {
          for (rv, 0, 3) {
            tensor[((((ax0.ax1.fused*12) + ax2)*11) + ax3)] = max(tensor[((((ax0.ax1.fused*12) + ax2)*11) + ax3)], tvm_if_then_else(((((3 - rv) <= (ax2*3)) && ((ax2*3) < (34 - rv))) && ((2 - rv) <= (ax3*3))), A[((((((ax0.ax1.fused*961) + (ax2*93)) + (rv*31)) + (ax3*3)) + rv) + -95)], -340282346638528859811704183484516925440.000000f))
          }
        }
      }
    }
  }
}
produce compute {
  for (i1, 0, 256) {
    for (i2, 0, 12) {
      for (i3, 0, 11) {
        compute[((((i1*12) + i2)*11) + i3)] = max(tensor[((((i1*12) + i2)*11) + i3)], 0.000000f)
      }
    }
  }
}
```
And for adaptive pooling:
```
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
produce tensor {
  parallel (ax0.ax1.fused, 0, 3) {
    for (ax2, 0, 100) {
      for (ax3, 0, 100) {
        tensor[((((ax0.ax1.fused*100) + ax2)*100) + ax3)] = 0.000000f
        for (rv, 0, (ceil((((ax2*224) + 224)/100)) - floor(((ax2*224)/100)))) {
          for (rv, 0, (ceil((((ax3*224) + 224)/100)) - floor(((ax3*224)/100)))) {
            if (likely((rv < (ceil((((ax2*224) + 224)/100)) - floor(((ax2*224)/100)))))) {
              if (likely((rv < (ceil((((ax3*224) + 224)/100)) - floor(((ax3*224)/100)))))) {
                tensor[((((ax0.ax1.fused*100) + ax2)*100) + ax3)] = (tensor[((((ax0.ax1.fused*100) + ax2)*100) + ax3)] + (A[((((((ax0.ax1.fused*224) + rv) + floor(((ax2*224)/100)))*224) + rv) + floor(((ax3*224)/100)))]/float32(((ceil((((ax2*224) + 224)/100)) - floor(((ax2*224)/100)))*(ceil((((ax3*224) + 224)/100)) - floor(((ax3*224)/100)))))))
              }
            }
          }
        }
      }
    }
  }
}

[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
[21:31:38] /home/ubuntu/tvm/src/arithmetic/int_set.cc:559: cannot evaluate set type Call
Traceback (most recent call last):
  File "test_topi_pooling.py", line 207, in <module>
    test_adaptive_pool()
  File "test_topi_pooling.py", line 196, in test_adaptive_pool
    verify_adaptive_pool(1, 3, 224, 224, [100, 100], 'avg')
  File "test_topi_pooling.py", line 193, in verify_adaptive_pool
    check_device(device)
  File "test_topi_pooling.py", line 188, in check_device
    f = tvm.build(s, [A, B], device)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/tvm-0.6.dev0-py3.7-linux-x86_64.egg/tvm/build_module.py", line 569, in build
    binds=binds)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/tvm-0.6.dev0-py3.7-linux-x86_64.egg/tvm/build_module.py", line 412, in lower
    return ir_pass.MakeAPI(stmt, name, arg_list, 0, cfg.restricted_func)
  File "/home/ubuntu/.local/lib/python3.7/site-packages/tvm-0.6.dev0-py3.7-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 206, in __call__
    raise get_last_ffi_error()
tvm._ffi.base.TVMError: Traceback (most recent call last):
  [bt] (3) /home/ubuntu/.local/lib/python3.7/site-packages/tvm-0.6.dev0-py3.7-linux-x86_64.egg/tvm/libtvm.so(TVMFuncCall+0x61) [0x7ff4e9cf1e81]
  [bt] (2) /home/ubuntu/.local/lib/python3.7/site-packages/tvm-0.6.dev0-py3.7-linux-x86_64.egg/tvm/libtvm.so(+0x17413b) [0x7ff4e95b813b]
  [bt] (1) /home/ubuntu/.local/lib/python3.7/site-packages/tvm-0.6.dev0-py3.7-linux-x86_64.egg/tvm/libtvm.so(+0x3bd75f) [0x7ff4e980175f]
  [bt] (0) /home/ubuntu/.local/lib/python3.7/site-packages/tvm-0.6.dev0-py3.7-linux-x86_64.egg/tvm/libtvm.so(+0x13ca92) [0x7ff4e9580a92]
  File "/home/ubuntu/tvm/src/pass/make_api.cc", line 188
TVMError: Not all Vars are passed in api_args:  'ax2'  'ax3'  does not appeared in api_args
```
@Laurawly @icemelon9 @yuruofeifei

Reply:
Error message for max pooling (previous comment is average pooling):
```
Traceback (most recent call last):
  File "test_topi_pooling.py", line 187, in <module>
    test_adaptive_pool()
  File "test_topi_pooling.py", line 176, in test_adaptive_pool
    verify_adaptive_pool(1, 3, 224, 224, [100, 100], 'max')
  File "test_topi_pooling.py", line 156, in verify_adaptive_pool
    b_np[i, j, k, l] = op(a_np[i, j, k_sl, l_sl])
  File "/home/ubuntu/miniconda3/envs/pt_latest/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 2334, in amax
    initial=initial)
  File "/home/ubuntu/miniconda3/envs/pt_latest/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 83, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
ValueError: zero-size array to reduction operation maximum which has no identity
```